internal/cmd/local/docker/docker.go

```diff
@@ -123,7 +123,9 @@ func newWithOptions(ctx context.Context, newPing newPing, goos string) (*Docker,
 
 // createAndPing attempts to create a docker client and ping it to ensure we can communicate
 func createAndPing(ctx context.Context, newPing newPing, host string, opts []client.Opt) (Client, error) {
-	cli, err := newPing(append(opts, client.WithHost(host))...)
+	// Pass client.WithHost first to ensure it runs prior to the client.FromEnv call.
+	// We want the DOCKER_HOST to be used if it has been specified, overriding our host.
+	cli, err := newPing(append([]client.Opt{client.WithHost(host)}, opts...)...)
```

Comment:
This seems reasonable. I'm guessing you verified that the order makes a difference. The only prior art I have for this looks like I have the same issue, so I can fix mine to do this

Reply:
the options are applied in a for-loop
```go
	for _, op := range ops {
		if err := op(c); err != nil {
			return nil, err
		}
	}
```
`WithHost` always sets the host (and related fields)
```go
func WithHost(host string) Opt {
	return func(c *Client) error {
		hostURL, err := ParseHostURL(host)
		if err != nil {
			return err
		}
		c.host = host
		c.proto = hostURL.Scheme
		c.addr = hostURL.Host
		c.basePath = hostURL.Path
		if transport, ok := c.client.Transport.(*http.Transport); ok {
			return sockets.ConfigureTransport(transport, c.proto, c.addr)
		}
		return errors.Errorf("cannot apply host to transport: %T", c.client.Transport)
	}
}
```
`WithHostFromEnv` (called from `FromEnv`) will call `WithHost` only if `DOCKER_HOST` is defined
```go
func WithHostFromEnv() Opt {
	return func(c *Client) error {
		if host := os.Getenv(EnvOverrideHost); host != "" {
			return WithHost(host)(c)
		}
		return nil
	}
}
```

Reply:
In the previous implementation, our `WithHost` was always overriding the `DOCKER_HOST`, now it's flipped so that `DOCKER_HOST` will override our `WithHost` (if `DOCKER_HOST` is defined).